docs/VirtualMachine.md

```diff
@@ -0,0 +1,160 @@
+
+#### Output
+
+An output consumes a stream and send the data to an external system.
+
+Output are reprented as functions that consume a stream and produce a stream.
+Outputs are transparent proxies, passing their input data downstream unmodified.
+As a result outputs themselves have no effect on the Flux program, in other words they only have side effects.
+
+#### Transformation
+
+A transformation is not a data type but rather an operation defined on a stream.
+A transformation consumes a stream of tables and produces a new stream of tables.
+
+Transformations may make use of a data cache to allow for multiple edits to a table before materializing the table into the new stream.
+
+Transformations are reprented as functions that consume a stream and produce a stream.
+
+#### Data cache
+
+A data cache is a set of tables that have yet to be materialized.
+Tables in a data cache may receive edits.
+The cache identifies each table by its group key.
+
+#### Triggers
+
+A trigger indicates that a table in the data cache should be materialized and delivered downstream.
+
+A trigger can fire multiple times until it is finished.
+Once a trigger is finished it will never fire again.
+
+A data cache uses triggers to know when it can evit and materialize tables.
+
+Triggers fire based on events.
+For example a trigger may fire after receiving an update watermark event.
+See the Flux spec for complete details on events available to triggers.
+
+#### Message bus
+
+The message bus is a queue where trigger events and data messages can be placed.
+Each transformation has its own message bus.
+A transformation consumes messages off its bus and processes them.
+
+### Data flow
+
+With the various data types and operations defined lets be explicit about how data flows through a Flux program.
+
+A Flux program describes a graph through which data flows.
+This graph may have complex data dependencies between transformations.
+Sources produce an inital stream of tables, each table is placed on the message bus for the downstream transformations of the source.
+Each transformation is linked to its downstream transformations via the message bus.
+Sources produces trigger events about the stream which are also placed on the message bus.
+The trigger events cause tables to be materialized from the data cache, if it exists. 
+The materialized tables are placed on the message bus of the downstream transformations.
+If a transformation does not need to make edits to a table it does not need to use the data cache but can simply place its generated tables directly on the downstream message bus.
+
+## Compiler
+
+The compiler is responsible for transforming a semantic graph into an IR.
+Internally the compiler performs several passes in order to produce optimized IR.
+
+1. Initial Pass - Convert the semantic graph into non-optimized IR.
+2. Planner - The planner rewrites the IR to optimize for the structure of the Flux script.
```

Comment:
I'm thinking #2 and #3 would ideally be written in Flux (or maybe in the IR?). #1 could still be in Go.